crates/stacks/src/docker.rs

```diff
@@ -0,0 +1,414 @@
+    container_name: String,
+    container_port: u16,
+    docker_bin: Option<&'static str>,
+    socket_path: Option<PathBuf>,
+    _state: PhantomData<State>,
+}
+
+impl<State> DockerManager<State> {
+    fn transition<NewState>(self) -> DockerManager<NewState> {
+        DockerManager {
+            data_dir: self.data_dir,
+            image_name: self.image_name,
+            container_name: self.container_name,
+            container_port: self.container_port,
+            docker_bin: self.docker_bin,
+            socket_path: self.socket_path,
+            _state: PhantomData,
+        }
+    }
+
+    fn docker_bin(&self) -> color_eyre::Result<&str> {
+        self.docker_bin
+            .ok_or_else(|| eyre!("Docker binary not set"))
+    }
+
+    fn socket_path(&self) -> color_eyre::Result<&PathBuf> {
+        self.socket_path
+            .as_ref()
+            .ok_or_else(|| eyre!("Socket path not set"))
+    }
+
+    fn is_container_running(&self, port: u16) -> color_eyre::Result<bool> {
+        let docker_bin = self.docker_bin()?;
+        let output = Command::new(docker_bin)
+            .args([
+                "ps",
+                "--filter",
+                &format!("name={}", self.container_name),
+                "--format",
+                "{{.Ports}}",
+            ])
+            .output()?;
+        let expected = format!("{port}->{}", self.container_port);
+        Ok(String::from_utf8_lossy(&output.stdout)
+            .trim()
+            .contains(&expected))
+    }
+}
+
+impl DockerManager<Initial> {
+    pub fn new(
+        data_dir: PathBuf,
+        image_name: String,
+        container_name: String,
+        container_port: u16,
+    ) -> Self {
+        Self {
+            data_dir,
+            image_name,
+            container_name,
+            container_port,
+            docker_bin: None,
+            socket_path: None,
+            _state: PhantomData,
+        }
+    }
+
+    pub fn check_socket_and_bin(mut self) -> color_eyre::Result<DockerManager<DockerSocketReady>> {
+        if Command::new("podman")
+            .arg("--version")
+            .output()
+            .map(|o| o.status.success())
+            .unwrap_or(false)
+        {
+            let xdg_runtime_dir =
+                env::var("XDG_RUNTIME_DIR").unwrap_or_else(|_| "/run/user/1000".to_string());
+            let podman_sock = PathBuf::from(format!("{xdg_runtime_dir}/podman/podman.sock"));
+            self.docker_bin = Some("podman");
+            self.socket_path = Some(podman_sock);
+        } else {
+            self.docker_bin = Some("docker");
+            self.socket_path = Some(PathBuf::from("/var/run/docker.sock"));
+        }
+        Ok(self.transition())
+    }
+}
+
+impl DockerManager<DockerSocketReady> {
+    pub fn check_docker_installed(self) -> color_eyre::Result<DockerManager<DockerInstalled>> {
+        let docker_bin = self.docker_bin()?;
+        info!("Checking for Docker/Podman...");
```

Comment:
feels like all of these info calls should be trace! only
or in some cases just removed